File: 383.Ransom_Note/main.go
package main

import "fmt"

type Solution struct{}

func (s Solution) canConstruct(ransomNote string, magazine string) bool {
    freq := make([]int, 26)
    for _, c := range magazine {
        freq[c - 'a']++
    }
    for _, c := range ransomNote {
        if freq[c - 'a']--; freq[c - 'a'] < 0 {
            return false
        }
    }
    return true
}

func main() {
    solution := Solution{}
    ransomNote := "aa"
    magazine := "aab"
    fmt.Println(solution.canConstruct(ransomNote, magazine)) // Output: true
}

File: 383.Ransom_Note/main.java
public class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] freq = new int[26];
        for (char c : magazine.toCharArray()) {
            freq[c - 'a']++;
        }
        for (char c : ransomNote.toCharArray()) {
            if (--freq[c - 'a'] < 0) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        String ransomNote = "aa";
        String magazine = "aab";
        System.out.println(solution.canConstruct(ransomNote, magazine)); // Output: true
    }
}
